### ISSUE TITLE
VOC training crash at the very begining on Windows

### ISSUE THREAD

**Issue by syyao98** [AlexeyAB/darknet#2124]
## Hardware
Intel i7-6700HQ
Nvidia 970M 6G
16G RAM
## Software
Windows 10
VS2015
CUDA10
cudnn 7.4.1
opencv 3.2
## Build option
Open `darknet.sln` -> `build` and successed(Release and x64)
## **Problem**
Followed this instruction (Train a Classifier on CIFAR-10)[https://pjreddie.com/darknet/train-cifar/], successed training the classifier. So `classifier.c` works well i guess.
But followed the VOC training instruction. Either AlexeyAB's or pjreddie's.
Training crashed at the very begining.
### Troubleshooting
I read some related issues and did some troubleshooting, couldn't locate the error.
`cfg/voc.data`
```
classes=20
train  = data/voc/train.txt
valid  = data/voc/2007_test.txt
names = data/voc.names
backup = backup/
```
- [x] UTF-8 and LF in all txt file, checked
- [x] Use absolute path in `cfg/voc.data`
- [x] Use a subset with 100 imgs of VOC
- [x] No empty line in txt file
- [x] Set all `random` to 0
- [x] Change `batch` and `subdivisions` to different combinations
- [x] No 0.0 bounding box
All failed, and still crash at the very begining
### Screenshot
followed the instruction, use pre-trained weight, crashed:
![image](https://user-images.githubusercontent.com/32408105/50533528-85bf7e80-0b66-11e9-87b4-dd377e30f3e2.png)
And the free space of my `C:` disk becomes 7G(from 20G)
![image](https://user-images.githubusercontent.com/32408105/50533567-637a3080-0b67-11e9-8410-e86416edf6dc.png)
Even no GPU usage, the training strucked
use the bash script(MinGW)
`./darknet.exe detector train cfg/voc.data cfg/yolov2-voc.cfg weights/darknet19_448.conv.23`
![image](https://user-images.githubusercontent.com/32408105/50533609-e00d0f00-0b67-11e9-8af6-494026bd9a86.png)
It crashed and threw me a segmentation fault.
## My Guess
I guess the there might be sth wrong when loading the data(failed at the 1st iteration)
`classifier.c` function well, and large net function well, i modify the darknet53, last conv filters: 1000 -> 10 and use it to train the cifar10 dataset, it works.
![image](https://user-images.githubusercontent.com/32408105/50533646-02ebf300-0b69-11e9-9b4c-b2f7bb62ac71.png)
And when it comes to `detector.c` it just crashed at the very begining.
And i notice that the training process just consumed my C disk memory then get stuck or crash with a segmentation fault.
So it may be something wrong when fetching the data or allocate the memory (or CUDA memory?) for the data?
Also `darknet.exe detector demo data/voc.data cfg/yolov2-voc.cfg weights/yolo-voc.weights` these test or demo scripts function well and gives a good result(using the official trained weights file).

**Comment by AlexeyAB** [comment]
@syyao98 Hi,

* Can you successfully get mAP (accuracy) for default trained model https://pjreddie.com/media/files/yolov2-voc.weights on Pascal VOC dataset? https://github.com/AlexeyAB/darknet#how-to-calculate-map-on-pascalvoc-2007

Using this command:
`darknet.exe detector map data/voc.data cfg/yolov2-voc.cfg yolo-voc.weights`


* Try to copy files to the same directory where is `darknet.exe`:

opencv_world340.dll, opencv_ffmpeg340_64.dll, 
cudnn64_7.dll, 
cublas64_100.dll, cudart64_100.dll, cufft64_100.dll, cusolver64_100.dll, cusparse64_100.dll
pthreadVC2.dll, pthreadGC2.dll, 


----

I can successfully train `yolov2_voc.cfg` on PascalVOC dataset:
`darknet.exe detector train data/voc.data cfg/yolov2-voc.cfg darknet19_448.conv.23 -dont_show`

* Windows 7 x64
* MSVS 2015
* CUDA 10.0 
* cuDNN 7.3.1 for CUDA 10.0
* OpenCV 3.3.0

File `data/voc.data`:
```
classes= 20
train  = data/train_voc.txt
valid  = data/2007_test.txt
#difficult = data/difficult_2007_test.txt
names = data/voc.names
backup = backup/
```

File `data/train_voc.txt`:
```
E:\VOC2007_2012\VOCtrainval_11-May-2012/VOCdevkit/VOC2007/JPEGImages/000012.jpg
E:\VOC2007_2012\VOCtrainval_11-May-2012/VOCdevkit/VOC2007/JPEGImages/000017.jpg
E:\VOC2007_2012\VOCtrainval_11-May-2012/VOCdevkit/VOC2007/JPEGImages/000023.jpg
E:\VOC2007_2012\VOCtrainval_11-May-2012/VOCdevkit/VOC2007/JPEGImages/000026.jpg
E:\VOC2007_2012\VOCtrainval_11-May-2012/VOCdevkit/VOC2007/JPEGImages/000032.jpg
E:\VOC2007_2012\VOCtrainval_11-May-2012/VOCdevkit/VOC2007/JPEGImages/000033.jpg
...
```

![image](https://user-images.githubusercontent.com/4096485/50538022-962d2480-0b79-11e9-9b4c-7b8b894ac736.png)

**Comment by syyao98** [comment]
## With nothing changed
Using the script `calc_mAP_voc_py.cmd`
![image](https://user-images.githubusercontent.com/32408105/50538323-bb855700-0ba8-11e9-8c4b-30d3738108e6.png)
It looks like the eval procedure function well
![image](https://user-images.githubusercontent.com/32408105/50538355-5d0ca880-0ba9-11e9-94b8-75d012cef962.png)
Using `darknet.exe detector map cfg/voc.data cfg/yolov2-voc.cfg weights/yolo-voc.weights`
(my weights file all stored at x64/weights and i use cfg/voc.data instead of data/voc.data),
Get the result below:
![image](https://user-images.githubusercontent.com/32408105/50538475-5ed76b80-0bab-11e9-879a-1c642c709998.png)

My opencv version may be 3.2? I have `opencv_ffmpeg320_64` and `opencv_world320.dll` with `darknet.exe`
![image](https://user-images.githubusercontent.com/32408105/50538414-82e67d00-0baa-11e9-878b-f83dc5836504.png)
I can use `darknet.exe detector demo` to use yolov2 or yolov3 weight to do the real time detection(about 12FPS), this may use opencv? So i guess opencv may not be the issue?
And i don't have  `cudnn64_7.dll`,`cublas64_100.dll`, `cudart64_100.dll`, `cufft64_100.dll`, `cusolver64_100.dll`, `cusparse64_100.dll` at the  root_dir ./x64, i just set the environment variable for the CUDA.
## With library files be copied
![image](https://user-images.githubusercontent.com/32408105/50538570-7bc06e80-0bac-11e9-97af-cfce19397d3a.png)
![image](https://user-images.githubusercontent.com/32408105/50538616-ebcef480-0bac-11e9-8696-7d479bc021ff.png)
It just crashed again... :cry:

**Comment by AlexeyAB** [comment]
@syyao98 This is very strange.

* Do you use the lates version from this repository?
* What is the message do you get each time? 

![image](https://user-images.githubusercontent.com/4096485/50539609-342de880-0b94-11e9-8920-e8627e9f1131.png)

* Did you change anything in the source code?

----

1. Try to set `train=` the same as the current `valid=` in the `voc.data` file:
```
train  = data/voc/2007_test.txt
valid  = data/voc/2007_test.txt
```
And try to train.

2. If it doesn't help - try to compile Darknet without OpenCV:
Open `darknet.sln` in the MSVS, (right click on project) -> properties -> C/C++ -> Preprocessor -> Preprocessor Definitions

Then remove here `OPENCV;`

Recompile.

And try to train.

3. If it doesn't help, try to compile `darknet_no_gpu.sln`
And try to train.

Do these 3 steps, and we will see where is a problem (in training dataset, OpenCV, CUDA/GPU, something else...).

**Comment by syyao98** [comment]
@AlexeyAB 
First, thank you for the reply!
I am sure that i use the latest version of this repository and didn't change anything in `darknet/src` and after build i modify some cfg files to train. 
So far the situation is:
- [x] Train a samll classifier on CIFAR10, using absolute path, relative path or symbolink dir, all successed
- [x] Train a large classifier on CIFAR10(modified darknet53), successed
- [x] Use trained weight to do the real time detection via cam(opencv API?), successed
- [x] Use the trained weight to do the evaluation, successed
- [ ] Use the same `.data` file, modify the `cfg` file(batch things), failed on VOC(either on pre-trained weight or from the scratch)
- [ ] Failed on a very small subset of VOC
- [ ] Failed on COCO training
 Very strange and almost drive me mad :anger:
Since evaluation passed while training failed, It seems the detector can't deal with batch that grater than 1?
Note that i have MinGW installed
With the windows command prompt, it just crashed with no error information
![image](https://user-images.githubusercontent.com/32408105/50539821-d7e1bd80-0bc1-11e9-9988-4315fd61cc07.png)
But in git bash
![image](https://user-images.githubusercontent.com/32408105/50539833-0495d500-0bc2-11e9-8c03-af9ea4125119.png)
It says i got a **segmentation fault** at the Line1. Since i only have 1 line in my bash script and it's the line to do the `./darknet detector train` thing. I guess i get a segmentation fault when fetching batches of data.(And she didn't tell me that and just crashed again and agian.
## As you adviced
1. Set `train=` the same as the current `valid=` in the `voc.data` file:
```
train = data/voc/2007_test.txt
valid  = data/voc/2007_test.txt
```
And in `cfg` file:
```
[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=64
```
Train: `darknet.exe detector train cfg/voc.data cfg/yolov2-voc.cfg -dont_show`
Crashed after printing the Learning Rate(as usual)
And with `cfg` file modified:
```
batch=1
subdivisions=1
```
Do the same training again, gives me the error:
![image](https://user-images.githubusercontent.com/32408105/50539959-e7620600-0bc3-11e9-85de-827e5cf8ed1c.png)
And then get stuck(So it may be sth wroing in the `train_function()` in `detector.c`?)
And after pressing `CTRL+C`:
![image](https://user-images.githubusercontent.com/32408105/50539987-66efd500-0bc4-11e9-9ffc-08fc67f27671.png)
The command prompt just shows me the Learning Rate things then quit.
Then do the evaluation with nothing changed in the same prompt winddows
`darknet.exe detector map cfg/voc.data cfg/yolov2-voc.cfg weights/yolo-voc.weights`
![image](https://user-images.githubusercontent.com/32408105/50539998-9ef71800-0bc4-11e9-88e0-005ded74e786.png)
Funtion well as before.

And i'm doing step 2 and step 3 now, after that i will edit this comment

**Comment by AlexeyAB** [comment]
@syyao98 
When did you download your Darknet code? Try to download the latest version.
Also I tested Darknet on GCC and MSVS, but I didn't test it on MinGW.

**Comment by syyao98** [comment]
> @syyao98
> When did you download your Darknet code? Try to download the latest version.
> Also I tested Darknet on GCC and MSVS, but I didn't test it on MinGW.

It's the opencv issue!
Update my last comment later!

**Comment by syyao98** [comment]
> @syyao98
> When did you download your Darknet code? Try to download the latest version.
> Also I tested Darknet on GCC and MSVS, but I didn't test it on MinGW.

I use MinGW for convenience, like `wget` `tar` `cat`
And i can run things like `get_coco_dataset.sh` directly on Windows
Everthing over was build on MSVS2015

BTW, your font of your command prompt is awesome. May i ask what is that?
